File: src/main/java/mods/eln/sim/ThermalLoadInitializerByPowerDrop.java
package mods.eln.sim;

import mods.eln.Eln;

public class ThermalLoadInitializerByPowerDrop {

    public double maximumTemperature, minimumTemperature;
    double heatingTao;
    double TConductivityDrop;

    public double Rs;
    public double Rp;
    /**
     * Thermal capacitance.
     */
    public double C;

    /**
     * @param maximumTemperature Intended maximum temperature in celsius.
     * @param minimumTemperature Intended minimum temperature in celsius.
     * @param heatingTao
     * @param TConductivityDrop
     */
    public ThermalLoadInitializerByPowerDrop(double maximumTemperature, double minimumTemperature, double heatingTao, double TConductivityDrop) {
        this.TConductivityDrop = TConductivityDrop;
        this.minimumTemperature = minimumTemperature;
        this.heatingTao = heatingTao;
        this.maximumTemperature = maximumTemperature;
    }

    public void setMaximalPower(double power) {
        C = power * heatingTao / maximumTemperature;
        Rp = maximumTemperature / power;
        Rs = TConductivityDrop / power / 2;

        Eln.simulator.checkThermalLoad(Rs, Rp, C);
    }

    public void applyToThermalLoad(ThermalLoad load) {
        load.set(Rs, Rp, C);
    }

    public ThermalLoadInitializerByPowerDrop copy() {
        ThermalLoadInitializerByPowerDrop thermalLoad = new ThermalLoadInitializerByPowerDrop(maximumTemperature, minimumTemperature, heatingTao, TConductivityDrop);
        thermalLoad.Rp = Rp;
        thermalLoad.Rs = Rs;
        thermalLoad.C = C;
        return thermalLoad;
    }
}

File: src/main/java/mods/eln/sim/thermal/ThermalLoadInitializerByPowerDrop.kt
package mods.eln.sim.thermal

class ThermalLoadInitializerByPowerDrop(var warmLimit: Double, var coolLimit: Double, internal var heatingTao: Double, internal var TConductivityDrop: Double) {

    var Rs: Double = 0.toDouble()
    var Rp: Double = 0.toDouble()
    var C: Double = 0.toDouble()

    fun setMaximalPower(P: Double) {
        C = P * heatingTao / warmLimit
        Rp = warmLimit / P
        Rs = TConductivityDrop / P / 2.0

        ThermalLoad.checkThermalLoad(Rs, Rp, C)
    }

    fun applyTo(load: ThermalLoad) {
        load.set(Rs, Rp, C)
    }

    fun copy(): ThermalLoadInitializerByPowerDrop {
        val thermalLoad = ThermalLoadInitializerByPowerDrop(warmLimit, coolLimit, heatingTao, TConductivityDrop)
        thermalLoad.Rp = Rp
        thermalLoad.Rs = Rs
        thermalLoad.C = C
        return thermalLoad
    }
}
